File: src-python/core/services/recalc_state.py
# Copyright (c) 2026 FicForge Contributors
# Licensed under the GNU Affero General Public License v3.0.
# See LICENSE file in the project root for full license text.

"""重算全局状态。参见 PRD §4.3。

手动触发全量重建 characters_last_seen / last_scene_ending / last_confirmed_chapter_focus。
使用场景：外部工具修改章节文件后一键修复状态。
"""

from __future__ import annotations

import logging
from dataclasses import asdict
from pathlib import Path
from typing import Any

import frontmatter

from core.domain.character_scanner import scan_characters_in_chapter
from core.domain.text_utils import extract_last_scene_ending

logger = logging.getLogger(__name__)


def recalc_state(
    au_path: Path,
    state_repo: Any,
    chapter_repo: Any,
    project_repo: Any,
    fact_repo: Any = None,
) -> dict[str, Any]:
    """重算全局状态，返回重建结果。

    Args:
        au_path: AU 根目录。
        state_repo: StateRepository 实例。
        chapter_repo: ChapterRepository 实例。
        project_repo: ProjectRepository 实例。

    Returns:
        重建结果 dict。
    """
    au_id = str(au_path)

    try:
        state = state_repo.get(au_id)
    except Exception:
        # state.yaml 损坏或缺失时创建默认 state
        from core.domain.state import State
        state = State(au_id=au_id)

    # 读取 cast_registry
    try:
        project = project_repo.get(au_id)
        cast_registry = asdict(project.cast_registry) if project.cast_registry else {"characters": []}
    except Exception:
        cast_registry = {"characters": []}

    # 获取所有已确认章节
    try:
        chapters = chapter_repo.list_main(au_id)
    except Exception:
        chapters = []

    chapters_scanned = 0
    new_characters_last_seen: dict[str, int] = {}
    new_last_scene_ending = ""
    new_last_confirmed_focus: list[str] = []

    if not chapters:
        # current_chapter == 1 或无章节：全部清空
        state.characters_last_seen = {}
        state.last_scene_ending = ""
        state.last_confirmed_chapter_focus = []
        state.chapters_dirty = []
        state.chapter_focus = []
        state_repo.save(state)
        return {
            "characters_last_seen": {},
            "last_scene_ending": "",
            "last_confirmed_chapter_focus": [],
            "chapters_scanned": 0,
            "cleaned_dirty_count": 0,
            "cleaned_focus_count": 0,
        }

    # 按章节号排序
    sorted_chapters = sorted(chapters, key=lambda c: c.chapter_num)

    for ch in sorted_chapters:
        try:
            content = ch.content
            if not content:
                continue
        except Exception:
            continue

        chapters_scanned += 1

        # 扫描角色
        scanned = scan_characters_in_chapter(
            content, cast_registry, chapter_num=ch.chapter_num,
        )
        for char_name, ch_num in scanned.items():
            existing = new_characters_last_seen.get(char_name, 0)
            if ch_num > existing:
                new_characters_last_seen[char_name] = ch_num

    # 最后一章的信息
    last_chapter = sorted_chapters[-1]
    try:
        last_content = last_chapter.content or ""
        new_last_scene_ending = extract_last_scene_ending(last_content)
    except Exception:
        new_last_scene_ending = ""

    # 从最后一章读取 confirmed_focus
    new_last_confirmed_focus = list(getattr(last_chapter, "confirmed_focus", []) or [])

    # 清理 chapters_dirty：移除不存在章节文件的 dirty 标记
    existing_nums = {ch.chapter_num for ch in sorted_chapters}
    old_dirty = list(getattr(state, "chapters_dirty", []) or [])
    new_dirty = [n for n in old_dirty if n in existing_nums]
    cleaned_dirty_count = len(old_dirty) - len(new_dirty)

    # 清理 chapter_focus：移除失效的 fact_id（不存在或已 resolved）
    old_focus = list(getattr(state, "chapter_focus", []) or [])
    cleaned_focus_count = 0
    if old_focus and fact_repo:
        try:
            facts = fact_repo.list_all(au_id)
            valid_focus_ids = {f.id for f in facts if f.status == "unresolved"}
            new_focus = [fid for fid in old_focus if fid in valid_focus_ids]
            cleaned_focus_count = len(old_focus) - len(new_focus)
        except Exception:
            new_focus = old_focus
    else:
        new_focus = old_focus

    # 写回 state
    state.characters_last_seen = new_characters_last_seen
    state.last_scene_ending = new_last_scene_ending
    state.last_confirmed_chapter_focus = new_last_confirmed_focus
    state.chapters_dirty = new_dirty
    state.chapter_focus = new_focus
    state_repo.save(state)

    return {
        "characters_last_seen": new_characters_last_seen,
        "last_scene_ending": new_last_scene_ending,
        "last_confirmed_chapter_focus": new_last_confirmed_focus,
        "chapters_scanned": chapters_scanned,
        "cleaned_dirty_count": cleaned_dirty_count,
        "cleaned_focus_count": cleaned_focus_count,
    }

File: src-engine/services/recalc_state.ts
// Copyright (c) 2026 FicForge Contributors
// Licensed under the GNU Affero General Public License v3.0.

/**
 * 重算全局状态。参见 PRD §4.3。
 * 手动触发全量重建 characters_last_seen / last_scene_ending / last_confirmed_chapter_focus。
 */

import { mergeCharactersLastSeen, scanCharactersInChapter } from "../domain/character_scanner.js";
import { createState } from "../domain/state.js";
import { extractLastSceneEnding } from "../domain/text_utils.js";
import { logCatch } from "../logger/index.js";
import type { ChapterRepository } from "../repositories/interfaces/chapter.js";
import type { FactRepository } from "../repositories/interfaces/fact.js";
import type { ProjectRepository } from "../repositories/interfaces/project.js";
import type { StateRepository } from "../repositories/interfaces/state.js";

export interface RecalcResult {
  characters_last_seen: Record<string, number>;
  last_scene_ending: string;
  last_confirmed_chapter_focus: string[];
  chapters_scanned: number;
  cleaned_dirty_count: number;
  cleaned_focus_count: number;
  /** 计算后的 state 对象（调用方负责 save，以便 ops-first 写入）。 */
  state: Awaited<ReturnType<StateRepository["get"]>>;
}

export async function recalcState(
  au_id: string,
  state_repo: StateRepository,
  chapter_repo: ChapterRepository,
  project_repo: ProjectRepository,
  fact_repo?: FactRepository | null,
  // 别名归一化表（E8）：全量重扫必须与 confirm/undo 同表，否则「重算状态」会把
  // confirm 已按主名记录的 characters_last_seen 冲回别名盲状态（真回归）。null = 无表回退现状。
  character_aliases: Record<string, string[]> | null = null,
): Promise<RecalcResult> {
  // 读取 state
  let state: Awaited<ReturnType<StateRepository["get"]>>;
  try {
    state = await state_repo.get(au_id);
  } catch {
    state = createState({ au_id });
  }

  // 读取 cast_registry：project.yaml 缺失（null）或读失败都回退空名册（recalc 为
  // best-effort 修复工具，不因 project 读不出而中止），读失败落日志可诊断。
  let castRegistry: { characters?: string[] } = { characters: [] };
  try {
    const project = await project_repo.get(au_id);
    castRegistry = project?.cast_registry ?? { characters: [] };
  } catch (err) {
    logCatch("recalc", "project read failed; using empty cast_registry", err);
  }

  // 获取所有已确认章节
  let chapters: Awaited<ReturnType<ChapterRepository["list_main"]>>;
  try {
    chapters = await chapter_repo.list_main(au_id);
  } catch {
    chapters = [];
  }

  if (chapters.length === 0) {
    state.characters_last_seen = {};
    state.last_scene_ending = "";
    state.last_confirmed_chapter_focus = [];
    state.chapters_dirty = [];
    state.chapter_focus = [];
    return {
      characters_last_seen: {},
      last_scene_ending: "",
      last_confirmed_chapter_focus: [],
      chapters_scanned: 0,
      cleaned_dirty_count: 0,
      cleaned_focus_count: 0,
      state,
    };
  }

  // 按章节号排序
  const sortedChapters = [...chapters].sort((a, b) => a.chapter_num - b.chapter_num);

  let chaptersScanned = 0;
  const newCharactersLastSeen: Record<string, number> = {};

  for (const ch of sortedChapters) {
    if (!ch.content) continue;
    chaptersScanned++;

    const scanned = scanCharactersInChapter(ch.content, castRegistry, character_aliases, ch.chapter_num);
    mergeCharactersLastSeen(newCharactersLastSeen, scanned);
  }

  // 最后一章的信息
  const lastChapter = sortedChapters[sortedChapters.length - 1];
  const newLastSceneEnding = lastChapter.content ? extractLastSceneEnding(lastChapter.content) : "";
  const newLastConfirmedFocus = [...(lastChapter.confirmed_focus ?? [])];

  // 清理 chapters_dirty
  const existingNums = new Set(sortedChapters.map((ch) => ch.chapter_num));
  const oldDirty = [...(state.chapters_dirty ?? [])];
  const newDirty = oldDirty.filter((n) => existingNums.has(n));
  const cleanedDirtyCount = oldDirty.length - newDirty.length;

  // 清理 chapter_focus
  const oldFocus = [...(state.chapter_focus ?? [])];
  let cleanedFocusCount = 0;
  let newFocus = oldFocus;

  if (oldFocus.length > 0 && fact_repo) {
    try {
      const facts = await fact_repo.list_all(au_id);
      const validFocusIds = new Set(facts.filter((f) => f.status === "unresolved").map((f) => f.id));
      newFocus = oldFocus.filter((fid) => validFocusIds.has(fid));
      cleanedFocusCount = oldFocus.length - newFocus.length;
    } catch {
      newFocus = oldFocus;
    }
  }

  // 计算完毕，更新 state 对象（不落盘，由调用方在 ops 之后 save）
  state.characters_last_seen = newCharactersLastSeen;
  state.last_scene_ending = newLastSceneEnding;
  state.last_confirmed_chapter_focus = newLastConfirmedFocus;
  state.chapters_dirty = newDirty;
  state.chapter_focus = newFocus;

  return {
    characters_last_seen: newCharactersLastSeen,
    last_scene_ending: newLastSceneEnding,
    last_confirmed_chapter_focus: newLastConfirmedFocus,
    chapters_scanned: chaptersScanned,
    cleaned_dirty_count: cleanedDirtyCount,
    cleaned_focus_count: cleanedFocusCount,
    state,
  };
}
